File: 0-Arrays/1481-Least-Number-of-Unique-Integers-after-K-Removals/1481-Least-Number-of-Unique-Integers-after-K-Removals.cpp
class Solution {
public:
    int findLeastNumOfUniqueInts(vector<int>& arr, int k) {
        map<int, int> numCount;
        for (int i = 0; i < arr.size(); ++i)
        {
            numCount[arr[i]]++;
        }
        map<int, int> freqMap;
        for (map<int, int>::iterator itr = numCount.begin(); 
            itr != numCount.end(); ++itr)
        {
            freqMap[itr->second]++;
        }
        int counter = 0;
        for (map<int, int>::iterator itr = freqMap.begin(); 
            itr != freqMap.end(); ++itr)
        {
            while (itr->first <= k && itr->second > 0)
            {
                k = k - itr->first;
                itr->second--;
            }
            counter = counter + itr->second;
            
        }
        return counter;
    }
};

/**
 * Notes: Use bucket sort to arrange the frequency count of all the
 * numbers. Then based on the frequency, decrease k until it reaches
 * 0 or can't be decreased. Then the final count of unique numbers is
 * the answer. 
 * */

File: 0-Sort/1481-Least-Number-of-Unique-Integers-after-K-Removals/1481-Least-Number-of-Unique-Integers-after-K-Removals.java
class Solution {
    public int findLeastNumOfUniqueInts(int[] arr, int k) {
        Map<Integer, Integer> countMap= new HashMap<>();
        for (int i = 0; i < arr.length; ++i)
        {
            if (!countMap.containsKey(arr[i]))
                countMap.put(arr[i], 1);
            else
                countMap.replace(arr[i], countMap.get(arr[i]) + 1);
        }
        List<Integer> freq = new ArrayList<>();
        for (int i: countMap.keySet())
            freq.add(countMap.get(i));
        int ans = countMap.size();
        Collections.sort(freq);
        int i = 0;
        while (k > 0 && freq.get(i) <= k)
        {
            k = k - freq.get(i);
            i++;
            ans--;
        }
        return ans;
    }
}
/**
 * Notes: Since we don't need to know what the numbers are, we 
 * can simply sort all the frequency of different numbers and 
 * starting to decrease from the overall unique numbers count
 * to the point we can remove any more. 
 */